README.md
# Let's Encrypt: Boulder cookbook

This is a cookbook for provisioning [Boulder][], an
[ACME-based][acme-spec] certificate authority, written in Go. The
Boulder application is an official effort of [Let's Encrypt
project][letsencrypt]. This particular cookbook is a fork of
[letsencrypt-boulder-server](https://github.com/patcon/chef-letsencrypt-boulder-server)
for use at the OSUOSL.

**Warning:** This cookbook was created for testing other cookbooks, not
production purposes.

## Supported Platforms

* CentOS 6
* CentOS 7

## Attributes

|**Key**                                         | **Type**| **Description**                                    |
|------------------------------------------------|---------|----------------------------------------------------|
|`['boulder']['config']['boulder-config']`       | Hash    | Deep-merged into [eponymous config file][config1]. |
|`['boulder']['config']['issuer-ocsp-responder']`| Hash    | Deep-merged into [eponymous config file][config2]. |

## Recipes

### `default`

Install and starts the Boulder server.

## Notes

* Pay close attention to the output of any failed attempts. By default,
  Boulder tries to use all ports in the range 8000 to 8010, among
others. If there is a conflict with other applications, you can use this
cookbook's attributes to force Boulder onto a different port.

* If you're running the boulder server during testing, and it's on the
  same server as the webserver, you'll likely want to edit the
hostsfile. The [`hostfile` cookbook][hostsfile-ckbk] is great for this.

## Development

To tag and publish a new version of this cookbook, first ensure:

* the `metadata.rb` version has been bumped appropriately,
* the `CHANGELOG.md` has been updated, and
* all changes have been committed to git.

Once that has been done:

    bundle exec rake publish

This will create a git tag and push a new release to the Supermarket.

## License and Authors
* Author:: Thijs Houtenbos (@thoutenbos) <thoutenbos@schubergphilis.com>
* Countributor:: Patrick Connolly (@patcon) <patrickcconnolly@gmail.com>

## Sponsors

Code contributions have been generously made by the following
organizations:

* [Schuberg Philis][schubergphilis] (@thoutenbos)
* [Blended Perspectives, Inc.][blendive] (@patcon)

[acme-spec]: https://github.com/letsencrypt/acme-spec/
[letsencrypt]: https://letsencrypt.org/
[boulder]: https://github.com/letsencrypt/boulder/
[schubergphilis]: https://www.schubergphilis.com/
[blendive]: http://www.blendedperspectives.com/
[config1]: https://github.com/letsencrypt/boulder/blob/master/test/boulder-config.json
[config2]: https://github.com/letsencrypt/boulder/blob/master/test/issuer-ocsp-responder.json
[hostsfile-ckbk]: https://github.com/customink-webops/hostsfile


CHANGELOG.md
3.2.0 (2020-06-24)
------------------
- Chef 15 compatibility fixes

3.1.1 (2019-10-28)
------------------
- Update boulder to revision release-2019-10-28

3.1.0 (2019-09-16)
------------------
- Chef 14 compatibility fixes

3.0.1 (2018-09-07)
------------------
- Update boulder version

3.0.0 (2018-08-10)
------------------
- Chef 13 compatibility fixes

2.0.1 (2018-05-23)
------------------
- Switch to using go-1.8 on legacy setup

2.0.0 (2018-02-21)
------------------
- Switch to using docker environment for boulder server to reduce converge times

1.0.1 (2017-11-14)
------------------
- Upgrade go version to 1.7

1.0.0 (2017-06-26)
------------------
- Cleanup and add support for CentOS 6

# 0.1.2

* Added more specific platform version specification.
* Get boulder git revision from attribute

# 0.1.1

* Added metadata for Supermarket.

# 0.1.0

* Initial release of `letsencrypt-boulder-server`, based off
  [`acme_server` test cookbook](https://github.com/schubergphilis/letsencrypt/tree/master/test/fixtures/cookbooks/acme_server).


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

CENTOS_6 = {
  platform: 'centos',
  version: '6',
}.freeze

CENTOS_7 = {
  platform: 'centos',
  version: '7',
}.freeze

ALL_PLATFORMS = [
  CENTOS_7,
].freeze

RSpec.configure do |config|
  config.log_level = :warn
end


spec/unit/recipes/default_spec.rb
require_relative '../../spec_helper'

describe 'osl-letsencrypt-boulder-server::default' do
  context "#{CENTOS_6[:platform]} #{CENTOS_6[:version]}" do
    cached(:chef_run) do
      ChefSpec::SoloRunner.new(CENTOS_6).converge(described_recipe)
    end
    it 'raises an error' do
      expect { chef_run }.to raise_error(RuntimeError)
    end
  end
  ALL_PLATFORMS.each do |p|
    context "#{p[:platform]} #{p[:version]}" do
      cached(:chef_run) do
        ChefSpec::SoloRunner.new(p) do |node|
          node.normal['boulder']['host_aliases'] = %w(example.com foo.org)
        end.converge(described_recipe)
      end
      before do
        stub_command('/usr/local/bin/docker-compose ps -q | wc -l | grep 0').and_return(true)
        stub_command('screen -list boulder | /bin/grep 1\ Socket\ in')
      end
      it 'converges successfully' do
        expect { chef_run }.to_not raise_error
      end
      it do
        expect(chef_run).to install_package('dnsmasq, curl')
      end
      it do
        expect(chef_run).to create_template('/etc/dnsmasq.conf').with(source: 'dnsmasq.erb')
      end
      %w(example.com foo.org).each do |host|
        it do
          expect(chef_run).to render_file('/etc/dnsmasq.conf').with_content(%r{^address=/#{host}/10.0.0.2$})
        end
      end
      it do
        expect(chef_run.template('/etc/dnsmasq.conf')).to notify('service[dnsmasq]').to(:restart).immediately
      end
      it do
        expect(chef_run).to start_service('dnsmasq')
      end
      it do
        expect(chef_run).to enable_service('dnsmasq')
      end
      it do
        expect(chef_run).to install_chef_gem('rest-client').with(compile_time: false)
      end
      [
        /^domain example.org$/,
        /^search example.org$/,
        /^nameserver 127.0.0.1$/,
      ].each do |line|
        it do
          expect(chef_run).to render_file('/etc/resolv.conf').with_content(line)
        end
      end
      %w(boulder boulder-rabbitmq boulder-mysql).each do |host|
        it do
          expect(chef_run).to_not render_file('/etc/dnsmasq.conf').with_content(%r{^address=/#{host}/127.0.0.1$})
        end
      end
      it do
        expect(chef_run).to create_directory('/opt/boulder')
      end
      it do
        expect(chef_run).to checkout_git('/opt/boulder')
          .with(
            repository: 'https://github.com/letsencrypt/boulder',
            revision: 'release-2019-10-28'
          )
      end
      it do
        expect(chef_run).to create_docker_service('default').with(misc_opts: '--live-restore')
      end
      %w(boulder_config boulder_limit boulder_dns wait_for_bootstrap).each do |rb|
        it do
          expect(chef_run).to run_ruby_block(rb)
        end
      end
      it do
        expect(chef_run).to run_execute('/usr/local/bin/docker-compose up -d')
          .with(
            live_stream: true,
            cwd: '/opt/boulder'
          )
      end
    end
  end
end


test/smoke/default/default_test.rb
# # encoding: utf-8

# Inspec test for recipe osl-letsencrypt-boulder-server::default

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/


recipes/_legacy.rb
package %w(screen initscripts logrotate tar wget libtool-ltdl-devel)

node.default['mariadb']['use_default_repository'] = true
node.default['mariadb']['install']['version'] = '10.1'
node.default['go']['version'] = '1.8'

include_recipe 'mariadb::server'
include_recipe 'rabbitmq'
include_recipe 'golang'

start_cmd = ''
python_runtime '2.7'
start_cmd = 'scl enable python27'

boulderdir = node['boulder']['dir']

cookbook_file "#{boulderdir}/test/setup.sh" do
  mode 0755
end

ruby_block 'boulder_config' do
  block do
    node['boulder']['config'].keys.each do |filename|
      config = ::JSON.parse ::File.read "#{boulderdir}/test/#{filename}.json"
      ::File.write("#{boulderdir}/test/#{filename}.json.bak", ::JSON.pretty_generate(config))
      config = Chef::Mixin::DeepMerge.deep_merge(node['boulder']['config'][filename].to_hash, config)
      ::File.write("#{boulderdir}/test/#{filename}.json", ::JSON.pretty_generate(config))
    end
  end
end

ruby_block 'boulder_limit' do
  block do
    limit = ::YAML.load ::File.read "#{boulderdir}/test/rate-limit-policies.yml"
    limit['certificatesPerName']['threshold'] = 999
    limit['pendingAuthorizationsPerAccount']['threshold'] = 99
    ::File.write("#{boulderdir}/test/rate-limit-policies.yml", limit.to_yaml)
  end
end

bash 'boulder_setup' do
  live_stream true
  cwd boulderdir
  code 'source /etc/profile.d/golang.sh && GO15VENDOREXPERIMENT=1 ./test/setup.sh 2>&1 && touch setup.done'
  creates "#{boulderdir}/setup.done"
end

bash 'run_boulder' do
  live_stream true
  cwd boulderdir
  code "source /etc/profile.d/golang.sh && GO15VENDOREXPERIMENT=1 screen -LdmS boulder #{start_cmd} ./start.py"
  not_if 'screen -list boulder | /bin/grep 1\ Socket\ in'
end

ruby_block 'wait_for_bootstrap' do
  block do
    require 'rest-client'
    times = 0
    loop do
      times += 1
      begin
        client = RestClient.get 'http://127.0.0.1:4000/directory'
      rescue
        sleep 10
        puts ::File.read "#{boulderdir}/screenlog.0"
      end
      Chef::Application.fatal!('Failed to run boulder server') if times > 30
      break if client && client.code == 200
    end
  end
end


files/default/setup.sh
#!/bin/bash
#
# Fetch dependencies of Boulder that are necessary for development or testing,
# and configure database and RabbitMQ.
#

set -ev

go get \
  bitbucket.org/liamstask/goose/cmd/goose \
  github.com/golang/lint/golint \
  github.com/golang/mock/mockgen \
  github.com/golang/protobuf/proto \
  github.com/golang/protobuf/protoc-gen-go \
  github.com/jsha/listenbuddy \
  github.com/kisielk/errcheck \
  github.com/mattn/goveralls \
  github.com/modocache/gover \
  github.com/tools/godep \
  golang.org/x/tools/cmd/stringer \
  golang.org/x/tools/cover &

(curl -sL https://github.com/google/protobuf/releases/download/v2.6.1/protobuf-2.6.1.tar.gz | \
 tar -xzv &&
 cd protobuf-2.6.1 && ./configure --prefix=$HOME && make && make install) &

# Set up rabbitmq exchange
go run cmd/rabbitmq-setup/main.go -server amqp://boulder-rabbitmq &

# Wait for all the background commands to finish.
wait

# Create the database and roles
./test/create_db.sh


recipes/default.rb
#
# Author:: Thijs Houtenbos <thoutenbos@schubergphilis.com>
# Cookbook:: chef-letsencrypt-boulder-server
# Recipe:: default
#
# Copyright:: 2015-2020, Schuberg Philis
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# Keep containers up if the docker daemon is restarted

raise 'CentOS 6 is not supported!' if platform_family?('rhel') && node['platform_version'].to_i == 6

package %w(dnsmasq curl)

template '/etc/dnsmasq.conf' do
  source 'dnsmasq.erb'
  variables(
    hosts: node['boulder']['host_aliases']
  )
  notifies :restart, 'service[dnsmasq]', :immediately
end

service 'dnsmasq' do
  action [:start, :enable]
end

node.default['resolver']['domain'] = 'example.org'
node.default['resolver']['search'] = 'example.org'
node.default['resolver']['nameservers'] = %w(127.0.0.1)

include_recipe 'resolver'
include_recipe 'git'

build_essential 'default'

chef_gem 'rest-client' do
  action :install
  compile_time false
end

boulderdir = node['boulder']['dir']

directory boulderdir do
  recursive true
end

git boulderdir do
  repository 'https://github.com/letsencrypt/boulder'
  revision node['boulder']['revision']
  action :checkout
end

node.override['osl-docker']['service'] = { misc_opts: '--live-restore' }
include_recipe 'osl-docker'
include_recipe 'osl-docker::compose'

ruby_block 'boulder_config' do
  block do
    node['boulder']['config'].each_key do |filename|
      config = ::JSON.parse ::File.read "#{boulderdir}/test/config/#{filename}.json"
      ::File.write("#{boulderdir}/test/config/#{filename}.json.bak", ::JSON.pretty_generate(config))
      config = Chef::Mixin::DeepMerge.deep_merge(node['boulder']['config'][filename].to_hash, config)
      ::File.write("#{boulderdir}/test/config/#{filename}.json", ::JSON.pretty_generate(config))
    end
  end
end

ruby_block 'boulder_limit' do
  block do
    limit = ::YAML.load ::File.read "#{boulderdir}/test/rate-limit-policies.yml"
    limit['certificatesPerName']['threshold'] = 999
    limit['pendingAuthorizationsPerAccount']['threshold'] = 99
    ::File.write("#{boulderdir}/test/rate-limit-policies.yml", limit.to_yaml)
  end
end

ruby_block 'boulder_dns' do
  block do
    dns = ::YAML.load ::File.read "#{boulderdir}/docker-compose.yml"
    dns['services']['boulder']['environment']['FAKE_DNS'] = node['ipaddress']
    ::File.write("#{boulderdir}/docker-compose.yml", dns.to_yaml)
  end
end

execute '/usr/local/bin/docker-compose up -d' do
  live_stream true
  cwd boulderdir
  only_if '/usr/local/bin/docker-compose ps -q | wc -l | grep 0'
end

ruby_block 'wait_for_bootstrap' do
  block do
    require 'rest-client'
    times = 0
    loop do
      times += 1
      begin
        client = RestClient.get 'http://127.0.0.1:4000/directory'
      rescue
        sleep 10
        print "\nStill waiting for boulder to start.. #{times * 10} seconds"
      end
      raise('Failed to run boulder server') if times > 30
      break if client && client.code == 200
    end
  end
end
